packages/blade/src/components/Checkbox/Checkbox.tsx

```diff
@@ -0,0 +1,204 @@
+import { useCheckbox } from './useCheckbox';
+import { Wrapper } from './Wrapper';
+
+type CheckboxProps = {
+  /**
+   * If `true`, sets the checkbox will be checked.
+   * Use `onChange` to update its value
+   *
+   * @default false
+   */
+  isChecked?: boolean;
+  /**
+   * If `true`, the checkbox will be initially checked.
+   *
+   * @default false
+   */
+  defaultChecked?: boolean;
+  /**
+   * The callback invoked when the checked state of the `Checkbox` changes.
+   */
+  onChange?: (isChecked: boolean) => void;
+  /**
+   * Sets the label text of the checkbox
+   */
+  children: string;
+  /**
+   * Help text of the checkbox
+   */
+  helpText?: string;
+  /**
+   * Error text of the checkbox
+   *
+   * Renders when `validationState` is set to 'error'
+   *
+   * Overrides helpText
+   */
+  errorText?: string;
+  /**
+   * If `true`, the checkbox will be indeterminate.
+   * This does not modify the isChecked property.
+   *
+   * @default false
+   */
+  isIndeterminate?: boolean;
+  /**
+   * The name of the input field in a checkbox
+   * (Useful for form submission).
+   */
+  name?: string;
+  /**
+   * The value to be used in the checkbox input.
+   * This is the value that will be returned on form submission.
+   */
+  value?: string;
+  /**
+   * If `true`, the checkbox will be disabled
+   *
+   * @default false
+   */
+  isDisabled?: boolean;
+  /**
+   * If `true`, the checkbox input is marked as required,
+   * and `required` attribute will be added
+   *
+   * @default false
+   */
+  isRequired?: boolean;
+  /**
+   * If `error`, the checkbox input is marked as invalid,
+   * and `invalid` attribute will be added
+   *
+   * @default undefined
+   */
+  validationState?: 'error' | undefined;
+};
+
+const Checkbox = ({
+  defaultChecked,
+  validationState,
+  isChecked,
+  isDisabled,
+  isIndeterminate,
+  isRequired,
+  name,
+  onChange,
+  value,
+  children,
+  helpText,
+  errorText,
+}: CheckboxProps): React.ReactElement => {
+  const groupProps = useCheckboxGroupContext();
+
+  // ban controlling checkbox manually while inside group
+  const hasDefaultChecked = !isUndefined(defaultChecked);
+  const hasIsChecked = !isUndefined(isChecked);
+  const hasOnChange = !isUndefined(onChange);
+  if ((hasDefaultChecked || hasIsChecked || hasOnChange) && !isEmpty(groupProps)) {
+    const props = [
+      hasDefaultChecked ? 'defaultChecked' : undefined,
+      hasIsChecked ? 'isChecked' : undefined,
+      hasOnChange ? 'onChange' : undefined,
+    ].filter(Boolean);
+
+    throw new Error(
+      `[Blade Checkbox]: Cannot set \`${props.join(
+        ',',
+      )}\` on <Checkbox /> when it's inside <CheckboxGroup />, Please set it on the <CheckboxGroup /> itself`,
+    );
+  }
+  // mandate value prop when using inside group
+  if (!value && !isEmpty(groupProps)) {
+    throw new Error(
+      `[Blade Checkbox]: <CheckboxGroup /> requires that you pass unique "value" prop to each <Checkbox />
```

Comment:
why so? If i'm using checkbox doesn't necessarily mean i would always use it for form-level things right? so why is this required?


also, i think all the things for checkbox related to checboxgroup shall be part of checboxgroup file we can always iterate on the children and add these validations for props there. this is just polluting checkbox unnecessarily. imagine tomorrow we'll want to use the checkbox for toggle component we'll be spilling this logic for that component as well

Reply:
> why so? If i'm using checkboxgroup doesn't necessarily mean i would always use it for form-level things right? so why is this required?

While inside the Group the value prop doesn't just about the form level value it also needed for state management of the checkbox group. We need the value of the checkboxes otherwise we don't know which checkbox to control/manage. 

```ts
      <CheckboxGroup
        defaultValue={['this', 'is', 'mapped']}
      >
        <CheckboxComponent value="this">Apple</CheckboxComponent>
        <CheckboxComponent value="is">Mango</CheckboxComponent>
        <CheckboxComponent value="mapped">Orange</CheckboxComponent>
      </CheckboxGroup>
```

> also, i think all the things for checkbox related to checkboxgroup shall be part of checboxgroup file we can always iterate on the children and add these validations for props there

Unfortunately that's not a reliable way to do that. Iteration will first of all have a runtime hit. 
Plus we won't be able to do the validation while iterating, if for some reason consumer wraps something with our Checkbox which could very much be a case for form libraries like react-hook-form which needs third party form elements to be wrapped in their `<Controller />` component

While in other hand validating via the Context is sure shot a validation no matter what you do with the checkbox. And is the most cleanest way to detect if it's inside of the CheckboxGroup or not.